interpreter/cling/lib/Interpreter/Interpreter.cpp

```diff
@@ -257,6 +257,16 @@ namespace cling {
       return;
 
     if (!isInSyntaxOnlyMode()) {
+      // check if the debug flag was set
+      // if the flag was set, add GDBRegistrationListener to JIT instance
+      for (auto const& s : m_Opts.CompilerOpts.Remaining) {
+        if (std::strcmp(s, "-g") == 0) {
+          getCI()->getCodeGenOpts().setDebugInfo(
```

Comment:
Isn’t this automatically set by the CIFactory?

Reply:
@SimeonEhrig, ping.

Reply:
Sorry, I was pretty busy this month. Let's see, if I can find it out.

Reply:
I see `-debug-info-kind=standalone` which might be able to set it?

Reply:
Nope, if I run `cling -g -debug-info-kind=standalone`, the debug level is `clang::codegenoptions::DebugInfoConstructor (unsigned int) 4`

Reply:
Maybe we need a different value?

Reply:
The documentation clearly say, emit source code debug information: https://releases.llvm.org/13.0.0/tools/clang/docs/ClangCommandLineReference.html#debug-information-generation
And if I set `-gfull`, I expect that all possible debug information are generated.

I strongly expect, that the setting of the debug level is done or changed somewhere in in the cling code.

Reply:
I don't understand that code then https://github.com/llvm/llvm-project/blob/75e33f71c2dae584b13a7d1186ae0a038ba98838/clang/lib/Frontend/CompilerInvocation.cpp#L1623

I agree we might be resetting the codegen options somewhere which would be pretty bad.

It seems we have similar code here: https://github.com/root-project/root/blob/f0a240b8554d08f4f3269e02e000d86d10381946/interpreter/cling/lib/Interpreter/CIFactory.cpp#L1671-L1672

Reply:
I found out, where the concrete debug level is set and how to enable full debug information.

The concrete debug level is set via `clang -cc1 -debug-info-kind=constructor ...` argument. This means, the decision happens in the CI factory, when the variable `CC1Args` is created:

https://github.com/vgvassilev/root/blob/1941a034acc32ad0adbc51d333a6a4d60ccc6b5e/interpreter/cling/lib/Interpreter/CIFactory.cpp#L1387-L1399

This means the clang code is responsible to choose the internal debug level and our cling "user code" does not affect it. This line of code chooses the debug level, if I set `-g` or `-gfull`: https://github.com/vgvassilev/root/blob/1941a034acc32ad0adbc51d333a6a4d60ccc6b5e/interpreter/llvm/src/tools/clang/lib/Driver/ToolChains/Clang.cpp#L3917

After a text search I found also the only line of code, which can set `FullDebugInfo`: https://github.com/vgvassilev/root/blob/1941a034acc32ad0adbc51d333a6a4d60ccc6b5e/interpreter/llvm/src/tools/clang/lib/Driver/ToolChains/Clang.cpp#L4066

Thanks to comments I know, that I have to run `cling -g -fstandalone-debug` to enable full debug information.

Okay, how we want to solve it? I have three ideas:

1. Document, that the user needs to pass the arguments `cling -g -fstandalone-debug` and don't manipulate the debug level inside cling.
2. Keep the current solution
3. Add the argument `-fstandalone-debug` before parsing the cc1 arguments if `-g` or other debug flags are set.

Reply:
@vgvassilev ping

Reply:
I'd prefer option 1.) because when we go to re-use clang-repl interfaces we will just use it as it is. I was thinking that when we call `cling -g` we can add also `-fstandalone-debug`, however that would require parsing arguments which can be a bit ugly. Another option might be if we detect `CLING_DEBUG` env variable to be set we add both flags.

Reply:
Okay, sounds reasonable. Is the llvm 13 port already rebased on the cling documentation PR https://github.com/root-project/root/pull/11474 ? I like to document there.

Reply:
@vgvassilev I didn't introduced an environment variable to add the flags. This requires at least changes in two different places in the code, where I have to check for the env variable. So the direct connection of both checks is little bit hidden. I think the extra work to maintain is bigger than the benefit.

But I introduced a second environment variable.
* `CLING_JITLINK`: enables the `jitlink` linker
*  `CLING_JITLINK_DEBUG`: enables the `jitlink` linker and register the gdb listener